spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskService.java

```diff
@@ -171,75 +179,110 @@ public DefaultTaskService(DataSourceProperties dataSourceProperties,
-		if (taskDefinition == null) {
-			throw new NoSuchTaskDefinitionException(taskName);
-		}
-		TaskParser taskParser = new TaskParser(taskDefinition.getName(), taskDefinition.getDslText(), true, true);
-		TaskNode taskNode = taskParser.parse();
-		// if composed task definition replace definition with one composed task
-		// runner and executable graph.
-		if (taskNode.isComposed()) {
-			taskDefinition = new TaskDefinition(taskDefinition.getName(),
-					TaskServiceUtils.createComposedTaskDefinition(
-							taskNode.toExecutableDSL(), taskConfigurationProperties));
-			taskDeploymentProperties = TaskServiceUtils.establishComposedTaskProperties(taskDeploymentProperties,
-					taskNode);
-		}
-
-		AppRegistration appRegistration = this.registry.find(taskDefinition.getRegisteredAppName(),
-				ApplicationType.task);
-		Assert.notNull(appRegistration, "Unknown task app: " + taskDefinition.getRegisteredAppName());
-		Resource appResource = this.registry.getAppResource(appRegistration);
-		Resource metadataResource = this.registry.getAppMetadataResource(appRegistration);
-
-		TaskExecution taskExecution = taskExecutionRepository.createTaskExecution(taskName);
-		taskDefinition = TaskServiceUtils.updateTaskProperties(taskDefinition,
-				dataSourceProperties);
-
-		Map<String, String> appDeploymentProperties = new HashMap<>(commonApplicationProperties.getTask());
-		appDeploymentProperties.putAll(
-				TaskServiceUtils.extractAppProperties(taskDefinition.getRegisteredAppName(), taskDeploymentProperties));
-
-		Map<String, String> deployerDeploymentProperties = DeploymentPropertiesUtils
-				.extractAndQualifyDeployerProperties(taskDeploymentProperties, taskDefinition.getRegisteredAppName());
-		if (StringUtils.hasText(this.dataflowServerUri) && taskNode.isComposed()) {
-			TaskServiceUtils.updateDataFlowUriIfNeeded(this.dataflowServerUri, appDeploymentProperties,
-					commandLineArgs);
-		}
-		AppDefinition revisedDefinition = TaskServiceUtils.mergeAndExpandAppProperties(taskDefinition, metadataResource,
-				appDeploymentProperties, this.whitelistProperties);
-		List<String> updatedCmdLineArgs = this.updateCommandLineArgs(commandLineArgs, taskExecution);
-		AppDeploymentRequest request = new AppDeploymentRequest(revisedDefinition, appResource,
-				deployerDeploymentProperties, updatedCmdLineArgs);
-		String id = this.taskLauncher.launch(request);
+		String id = this.taskLauncher.launch(requestCriteria.getAppDeploymentRequest());
 		if (!StringUtils.hasText(id)) {
 			throw new IllegalStateException("Deployment ID is null for the task:" + taskName);
 		}
-		this.taskExecutionRepository.updateExternalExecutionId(taskExecution.getExecutionId(), id);
+		postLaunchProcessing(requestCriteria, id);
+		return requestCriteria.getTaskExecution().getExecutionId();
+	}
+
+	private RequestCriteria createLaunchRequest (final String taskName, final Map<String, String> taskDeploymentProperties,
```

Comment:
It seems like from the usage of the class `RequestCriteria` all the information can be obtained from the `AppDeploymentRequest`, so this method can change to `AppDeploymentRequest createLaunchRequest`

Reply:
RequestCriteria is still being used because we need the DSL in TaskDefinition that is not stored in AppDefinition stored in the AppDeploymentRequest.  Also while the executionId is available in the AppDefinition it is stored in the commandLineArgs and would have to be extracted via a search of the list to find it and then extract it from the string.  So storing it in the RequestCriteria appeared to be cleaner.